File: .skills/openclaw-skills/skills/0xnyk/xint/lib/x_search.ts
/**
 * lib/x_search.ts — xAI x_search integration
 *
 * Uses xAI's Responses API with the x_search tool to search X/Twitter
 * without needing cookies or GraphQL scraping.
 */

import { readFileSync } from "fs";
import { join } from "path";
import { trackCostDirect } from "./costs";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface XSearchResult {
  url?: string;
  tweet_url?: string;
  link?: string;
  text?: string;
  content?: string;
  snippet?: string;
  title?: string;
  username?: string;
  author?: string;
  handle?: string;
  created_at?: string;
  date?: string;
  timestamp?: string;
}

export interface Citation {
  url: string;
  title?: string;
  start_index?: number;
  end_index?: number;
}

export interface XSearchOptions {
  maxResults?: number;
  fromDate?: string;  // YYYY-MM-DD
  toDate?: string;   // YYYY-MM-DD
  model?: string;
  timeoutSeconds?: number;
  excludedDomains?: string[];
  allowedDomains?: string[];
  vision?: boolean;
}

export interface XSearchResponse {
  results: XSearchResult[];
  summary: string;
  citations: Citation[];
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const API_BASE = "https://api.x.ai/v1";
const DEFAULT_MODEL = "grok-4";
const DEFAULT_MAX_RESULTS = 10;
const DEFAULT_TIMEOUT = 45;

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

function getXaiKey(): string {
  if (process.env.XAI_API_KEY) return process.env.XAI_API_KEY;

  try {
    const envFile = readFileSync(join(import.meta.dir, "..", ".env"), "utf-8");
    const match = envFile.match(/XAI_API_KEY=["']?([^"'\n]+)/);
    if (match) return match[1];
  } catch {}

  throw new Error("XAI_API_KEY not found. Set it in your environment or in .env");
}

// ---------------------------------------------------------------------------
// Helper Functions
// ---------------------------------------------------------------------------

function bestUrl(r: XSearchResult): string {
  return r.url || r.tweet_url || r.link || "";
}

function bestText(r: XSearchResult): string {
  const text = r.text || r.content || r.snippet || r.title || "";
  return text.trim();
}

function bestHandle(r: XSearchResult): string {
  const handle = r.username || r.author || r.handle || "";
  return handle.replace(/^@/, "").trim();
}

function bestCreatedAt(r: XSearchResult): string {
  return r.created_at || r.date || r.timestamp || "";
}

// ---------------------------------------------------------------------------
// Core API
// ---------------------------------------------------------------------------

/**
 * Search X using xAI's hosted x_search tool.
 */
export async function xSearch(
  query: string,
  opts: XSearchOptions = {}
): Promise<XSearchResponse> {
  const apiKey = getXaiKey();
  const model = opts.model || DEFAULT_MODEL;
  const maxResults = opts.maxResults || DEFAULT_MAX_RESULTS;
  const timeout = (opts.timeoutSeconds || DEFAULT_TIMEOUT) * 1000;

  // Build tool spec
  const toolSpec: Record<string, unknown> = {
    type: "x_search",
    max_results: maxResults,
  };
  if (opts.fromDate) {
    toolSpec["from_date"] = opts.fromDate;
  }
  if (opts.toDate) {
    toolSpec["to_date"] = opts.toDate;
  }
  if (opts.excludedDomains && opts.excludedDomains.length > 0) {
    toolSpec["excluded_domains"] = opts.excludedDomains;
  }
  if (opts.allowedDomains && opts.allowedDomains.length > 0) {
    toolSpec["allowed_domains"] = opts.allowedDomains;
  }
  if (opts.vision) {
    toolSpec["enable_image_understanding"] = true;
  }

  const body = {
    model,
    input: `Use x_search to find recent posts relevant to: ${query}\nReturn the search results.`,
    tools: [toolSpec],
    tool_choice: "required",
    max_output_tokens: 800,
  };

  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), timeout);

  try {
    const res = await fetch(`${API_BASE}/responses`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(body),
      signal: controller.signal,
    });

    clearTimeout(timeoutId);

    if (res.status === 401) {
      throw new Error("xAI auth failed (401). Check your XAI_API_KEY.");
    }
    if (res.status === 402) {
      throw new Error("xAI payment required (402). Your account may be out of credits.");
    }
    if (res.status === 429) {
      throw new Error("xAI rate limited (429). Try again in a moment.");
    }
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`xAI API error (${res.status}): ${text.slice(0, 500)}`);
    }

    const data = await res.json() as {
      output?: Array<{
        type?: string;
        results?: XSearchResult[];
        content?: Array<{
          type?: string;
          text?: string;
          annotations?: Array<{
            type?: string;
            url?: string;
            title?: string;
            start_index?: number;
            end_index?: number;
          }>;
        }>;
      }>;
      usage?: { input_tokens: number; output_tokens: number };
    };

    // Track xAI x_search cost
    if (data.usage) {
      const inputCost = (data.usage.input_tokens / 1_000_000) * 3.0;   // grok-4 input rate
      const outputCost = (data.usage.output_tokens / 1_000_000) * 15.0; // grok-4 output rate
      trackCostDirect("xai_x_search", `${API_BASE}/responses`, inputCost + outputCost);
    } else {
      // Fallback: use estimated per-call rate
      trackCostDirect("xai_x_search", `${API_BASE}/responses`, 0.002);
    }

    // Extract results, summary, and citations
    const results: XSearchResult[] = [];
    const citations: Citation[] = [];
    let summary = "";

    const output = data.output || [];
    for (const item of output) {
      if (item.type === "x_search_call") {
        const r = item.results;
        if (Array.isArray(r)) {
          results.push(...r);
        }
      }

      if (item.type === "message") {
        const content = item.content || [];
        for (const part of content) {
          if (part.type === "output_text" && part.text) {
            summary = part.text.trim();
          }
          if (part.annotations && Array.isArray(part.annotations)) {
            for (const ann of part.annotations) {
              if (ann.url) {
                citations.push({
                  url: ann.url,
                  title: ann.title,
                  start_index: ann.start_index,
                  end_index: ann.end_index,
                });
              }
            }
          }
        }
      }
    }

    return { results, summary, citations };
  } catch (err) {
    clearTimeout(timeoutId);
    throw err;
  }
}

// ---------------------------------------------------------------------------
// CLI Handler
// ---------------------------------------------------------------------------

export async function cmdXSearch(args: string[]): Promise<void> {
  const queriesFile = args[0];
  const outJson = args[1];
  const outMd = args[2];
  const opts: XSearchOptions = {
    maxResults: 10,
    model: DEFAULT_MODEL,
    timeoutSeconds: DEFAULT_TIMEOUT,
  };

  // Parse remaining args
  for (let i = 3; i < args.length; i++) {
    const arg = args[i];
    switch (arg) {
      case "--max-results":
        opts.maxResults = parseInt(args[++i]) || 10;
        break;
      case "--model":
        opts.model = args[++i];
        break;
      case "--from-date":
        opts.fromDate = args[++i];
        break;
      case "--to-date":
        opts.toDate = args[++i];
        break;
      case "--exclude-domains":
        opts.excludedDomains = (args[++i] || "").split(",").map(d => d.trim()).filter(Boolean);
        break;
      case "--allow-domains":
        opts.allowedDomains = (args[++i] || "").split(",").map(d => d.trim()).filter(Boolean);
        break;
      case "--vision":
        opts.vision = true;
        break;
    }
  }

  if (!queriesFile) {
    console.error("Usage: xint x_search <queries_file> <out_json> <out_md> [options]");
    console.error("Options:");
    console.error("  --max-results N    Max results per query (default: 10)");
    console.error("  --model <name>     Model (default: grok-4)");
    console.error("  --from-date YYYY-MM-DD");
    console.error("  --to-date YYYY-MM-DD");
    console.error("  --exclude-domains  Comma-separated domains to exclude");
    console.error("  --allow-domains    Comma-separated domains to allow");
    console.error("  --vision           Enable image understanding");
    process.exit(1);
  }

  // Load queries
  let queries: string[];
  try {
    const content = await Bun.file(queriesFile).text();
    const parsed = JSON.parse(content);
    if (Array.isArray(parsed)) {
      queries = parsed.filter((q: unknown) => typeof q === "string" && q.trim());
    } else if (parsed.queries && Array.isArray(parsed.queries)) {
      queries = parsed.queries.filter((q: unknown) => typeof q === "string" && q.trim());
    } else {
      throw new Error("Invalid queries file format");
    }
  } catch (e: any) {
    console.error(`Error loading queries: ${e.message}`);
    process.exit(1);
  }

  if (queries.length === 0) {
    console.error("No queries provided.");
    process.exit(1);
  }

  const ts = new Date().toISOString().replace("Z", "Z");
  const perQuery: Array<{
    query: string;
    results: XSearchResult[];
    summary: string;
    citations: Citation[];
    error?: string;
  }> = [];
  let hadErrors = false;

  console.log(`Running x_search for ${queries.length} query(s)...`);

  for (const q of queries) {
    try {
      const { results, summary, citations } = await xSearch(q, opts);
      perQuery.push({ query: q, results, summary, citations });
      console.log(`  "${q}": ${results.length} results`);
    } catch (e: any) {
      hadErrors = true;
      perQuery.push({
        query: q,
        results: [],
        summary: "",
        citations: [],
        error: e.message.slice(0, 500),
      });
      console.error(`  "${q}": ERROR - ${e.message.slice(0, 100)}`);
    }
  }

  // Build JSON payload
  const jsonPayload = {
    timestamp: ts,
    model: opts.model,
    max_results: opts.maxResults,
    from_date: opts.fromDate,
    to_date: opts.toDate,
    queries: perQuery.map((pq) => ({
      query: pq.query,
      summary: pq.summary,
      results: pq.results.map((r) => ({
        url: bestUrl(r),
        text: bestText(r),
        username: bestHandle(r),
        created_at: bestCreatedAt(r),
      })),
      citations: pq.citations.length > 0 ? pq.citations : undefined,
      error: pq.error,
    })),
  };

  // Write JSON output
  if (outJson) {
    try {
      await Bun.write(outJson, JSON.stringify(jsonPayload, null, 2) + "\n");
      console.log(`JSON: ${outJson}`);
    } catch (e: any) {
      console.error(`Error writing JSON: ${e.message}`);
    }
  }

  // Write markdown report
  if (outMd) {
    const md = renderMarkdown(ts, queries, perQuery);
    try {
      await Bun.write(outMd, md);
      console.log(`Report: ${outMd}`);
    } catch (e: any) {
      console.error(`Error writing markdown: ${e.message}`);
    }
  }

  // Summary
  const totalResults = perQuery.reduce((sum, pq) => sum + pq.results.length, 0);
  const status = hadErrors && totalResults === 0 ? "FAIL" : hadErrors ? "PARTIAL" : "OK";
  console.log(`\nxAI X search: ${status} (${queries.length} queries, ${totalResults} results)`);
}

// ---------------------------------------------------------------------------
// Markdown Rendering
// ---------------------------------------------------------------------------

function renderMarkdown(
  ts: string,
  queries: string[],
  perQuery: Array<{
    query: string;
    results: XSearchResult[];
    summary: string;
    citations: Citation[];
    error?: string;
  }>
): string {
  const lines: string[] = [];
  lines.push("# xAI X Search Scan");
  lines.push("");
  lines.push(`- Timestamp (UTC): ${ts}`);
  lines.push(`- Queries: ${queries.length}`);
  lines.push("");

  // Find combined summary (prefer last non-empty)
  const combinedSummary = perQuery
    .slice()
    .reverse()
    .find((pq) => pq.summary.trim())?.summary || "";

  if (combinedSummary) {
    lines.push("## Summary");
    lines.push("");
    lines.push(combinedSummary.trim());
    lines.push("");
  }

  lines.push("## Results");
  lines.push("");

  for (const pq of perQuery) {
    lines.push(`### Query: \`${pq.query}\``);
    lines.push("");

    if (pq.error) {
      lines.push(`- ERROR: ${pq.error}`);
      lines.push("");
      continue;
    }

    if (pq.results.length === 0) {
      lines.push("- (no results)");
      lines.push("");
      continue;
    }

    for (const r of pq.results) {
      const handle = bestHandle(r);
      let text = bestText(r).replace(/\n/g, " ");
      if (text.length > 220) {
        text = text.slice(0, 217) + "...";
      }
      const url = bestUrl(r);
      const createdAt = bestCreatedAt(r);

      const prefix = handle ? `@${handle}: ` : "";
      const meta = createdAt ? ` (${createdAt})` : "";

      if (url) {
        lines.push(`- ${prefix}${text}${meta} ${url}`);
      } else {
        lines.push(`- ${prefix}${text}${meta}`);
      }
    }
    lines.push("");
  }

  // Collect all citations across queries
  const allCitations = perQuery.flatMap((pq) => pq.citations);
  if (allCitations.length > 0) {
    const seen = new Set<string>();
    const unique = allCitations.filter((c) => {
      if (seen.has(c.url)) return false;
      seen.add(c.url);
      return true;
    });

    lines.push("## Citations");
    lines.push("");
    for (let i = 0; i < unique.length; i++) {
      const c = unique[i];
      const label = c.title || c.url;
      lines.push(`${i + 1}. [${label}](${c.url})`);
    }
    lines.push("");
  }

  return lines.join("\n").trimEnd() + "\n";
}

// ---------------------------------------------------------------------------
// Utility Exports
// ---------------------------------------------------------------------------

export { bestUrl, bestText, bestHandle, bestCreatedAt };

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/x_search.rs
use anyhow::{bail, Result};
use std::collections::HashSet;
use std::path::PathBuf;

use crate::api::xai;
use crate::cli::XSearchArgs;
use crate::config::Config;

pub async fn run(args: &XSearchArgs, config: &Config) -> Result<()> {
    let api_key = config.require_xai_key()?;
    let http = reqwest::Client::new();

    // Load queries
    let queries = load_queries(&args.queries_file)?;
    if queries.is_empty() {
        bail!("No queries provided.");
    }

    let ts = chrono::Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    let model = &args.model;
    let max_results = args.max_results;
    let timeout = args.timeout_seconds as u64;
    let from_date = args.from_date.as_deref();
    let to_date = args.to_date.as_deref();
    let excluded_domains: Option<Vec<String>> = args.exclude_domains.as_ref().map(|s| {
        s.split(',')
            .map(|d| d.trim().to_string())
            .filter(|d| !d.is_empty())
            .collect()
    });
    let allowed_domains: Option<Vec<String>> = args.allow_domains.as_ref().map(|s| {
        s.split(',')
            .map(|d| d.trim().to_string())
            .filter(|d| !d.is_empty())
            .collect()
    });

    let mut per_query: Vec<QueryResult> = Vec::new();
    let mut had_errors = false;

    for q in &queries {
        match xai::x_search(
            &http,
            api_key,
            q,
            max_results,
            from_date,
            to_date,
            model,
            timeout,
            excluded_domains.as_deref(),
            allowed_domains.as_deref(),
            args.vision,
        )
        .await
        {
            Ok((results, summary, citations)) => {
                per_query.push(QueryResult {
                    query: q.clone(),
                    results,
                    summary,
                    citations,
                    error: None,
                });
            }
            Err(e) => {
                had_errors = true;
                let err_msg = format!("{e}");
                per_query.push(QueryResult {
                    query: q.clone(),
                    results: Vec::new(),
                    summary: String::new(),
                    citations: Vec::new(),
                    error: Some(err_msg[..err_msg.len().min(500)].to_string()),
                });
            }
        }
    }

    // Build JSON payload
    let json_payload = build_json_payload(&ts, model, max_results, from_date, to_date, &per_query);

    // Write JSON output
    let out_json = PathBuf::from(&args.out_json);
    if let Some(parent) = out_json.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(
        &out_json,
        serde_json::to_string_pretty(&json_payload)? + "\n",
    )?;

    // Write markdown report
    let combined_summary = per_query
        .iter()
        .rev()
        .find_map(|qr| {
            if qr.summary.is_empty() {
                None
            } else {
                Some(qr.summary.clone())
            }
        })
        .unwrap_or_default();

    let md = render_md(&ts, &queries, &per_query, &combined_summary);
    let out_md = PathBuf::from(&args.out_md);
    if let Some(parent) = out_md.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(&out_md, &md)?;

    // Print status
    let total_results: usize = per_query.iter().map(|qr| qr.results.len()).sum();
    let status = if had_errors && total_results == 0 {
        "FAIL"
    } else if had_errors {
        "PARTIAL"
    } else {
        "OK"
    };
    eprintln!(
        "xAI X search: {} ({} queries, {} results)",
        status,
        queries.len(),
        total_results
    );
    eprintln!("Report: {}", out_md.display());

    // Emit memory candidates if requested
    if args.emit_candidates {
        let workspace = PathBuf::from(&args.workspace);
        let existing = load_existing_sources(&workspace);
        let today = chrono::Utc::now().format("%Y-%m-%d").to_string();
        let cand_out = if args.candidates_out.is_empty() {
            workspace
                .join("memory")
                .join("candidates")
                .join(format!("x-search-{today}.jsonl"))
        } else {
            PathBuf::from(&args.candidates_out)
        };
        if let Some(parent) = cand_out.parent() {
            std::fs::create_dir_all(parent)?;
        }

        let mut new_lines = Vec::new();
        let mut seen = existing;
        for qr in &per_query {
            if qr.error.is_some() {
                continue;
            }
            for r in &qr.results {
                let url = r.best_url();
                let text = r.best_text().replace('\n', " ");
                let handle = r.best_handle();
                if text.is_empty() && url.is_empty() {
                    continue;
                }
                let source = if !url.is_empty() {
                    format!("x_search:{url}")
                } else {
                    use std::hash::{Hash, Hasher};
                    let mut hasher = std::collections::hash_map::DefaultHasher::new();
                    text.hash(&mut hasher);
                    format!("x_search:{}:{}", handle, hasher.finish())
                };
                if seen.contains(&source) {
                    continue;
                }

                let mut val = text.clone();
                if !handle.is_empty() {
                    val = format!("@{handle}: {val}");
                }
                if !url.is_empty() {
                    val = format!("{val} ({url})");
                }
                if val.len() > 900 {
                    val.truncate(900);
                }

                let item = serde_json::json!({
                    "type": "fact",
                    "value": val,
                    "confidence": 0.55,
                    "source": source,
                    "created_at": ts,
                });
                new_lines.push(serde_json::to_string(&item)?);
                seen.insert(source);
            }
        }

        if !new_lines.is_empty() {
            use std::io::Write;
            let mut f = std::fs::OpenOptions::new()
                .create(true)
                .append(true)
                .open(&cand_out)?;
            for line in &new_lines {
                writeln!(f, "{line}")?;
            }
        }
        eprintln!("New memory candidates: {}", new_lines.len());
        if !new_lines.is_empty() {
            eprintln!("Candidates appended: {}", cand_out.display());
        }
    }

    Ok(())
}

struct QueryResult {
    query: String,
    results: Vec<xai::XSearchResult>,
    summary: String,
    citations: Vec<xai::Citation>,
    error: Option<String>,
}

fn load_queries(path: &str) -> Result<Vec<String>> {
    let content = std::fs::read_to_string(path)?;
    let val: serde_json::Value = serde_json::from_str(&content)?;

    if let Some(arr) = val.get("queries").and_then(|v| v.as_array()) {
        Ok(arr
            .iter()
            .filter_map(|v| v.as_str().map(|s| s.trim().to_string()))
            .filter(|s| !s.is_empty())
            .collect())
    } else if let Some(arr) = val.as_array() {
        Ok(arr
            .iter()
            .filter_map(|v| v.as_str().map(|s| s.trim().to_string()))
            .filter(|s| !s.is_empty())
            .collect())
    } else {
        bail!("queries file must be a JSON array of strings or {{\"queries\": [...]}} object")
    }
}

fn build_json_payload(
    ts: &str,
    model: &str,
    max_results: u32,
    from_date: Option<&str>,
    to_date: Option<&str>,
    per_query: &[QueryResult],
) -> serde_json::Value {
    let queries: Vec<serde_json::Value> = per_query
        .iter()
        .map(|qr| {
            serde_json::json!({
                "query": qr.query,
                "summary": qr.summary,
                "results": qr.results.iter().map(|r| {
                    serde_json::json!({
                        "url": r.best_url(),
                        "text": r.best_text(),
                        "username": r.best_handle(),
                        "created_at": r.best_created_at(),
                    })
                }).collect::<Vec<_>>(),
                "citations": qr.citations,
                "error": qr.error,
            })
        })
        .collect();

    serde_json::json!({
        "timestamp": ts,
        "model": model,
        "max_results": max_results,
        "from_date": from_date,
        "to_date": to_date,
        "queries": queries,
    })
}

fn render_md(ts: &str, queries: &[String], per_query: &[QueryResult], summary: &str) -> String {
    let mut lines = Vec::new();
    lines.push("# xAI X Search Scan".to_string());
    lines.push(String::new());
    lines.push(format!("- Timestamp (UTC): {ts}"));
    lines.push(format!("- Queries: {}", queries.len()));
    lines.push(String::new());

    if !summary.is_empty() {
        lines.push("## Summary".to_string());
        lines.push(String::new());
        lines.push(summary.trim().to_string());
        lines.push(String::new());
    }

    lines.push("## Results".to_string());
    lines.push(String::new());

    for qr in per_query {
        lines.push(format!("### Query: `{}`", qr.query));
        lines.push(String::new());
        if let Some(err) = &qr.error {
            lines.push(format!("- ERROR: {err}"));
            lines.push(String::new());
            continue;
        }
        if qr.results.is_empty() {
            lines.push("- (no results)".to_string());
            lines.push(String::new());
            continue;
        }
        for r in &qr.results {
            let handle = r.best_handle();
            let mut text = r.best_text().replace('\n', " ");
            if text.len() > 220 {
                text.truncate(217);
                text.push_str("...");
            }
            let url = r.best_url();
            let created_at = r.best_created_at();

            let prefix = if handle.is_empty() {
                String::new()
            } else {
                format!("@{handle}: ")
            };
            let meta = if created_at.is_empty() {
                String::new()
            } else {
                format!(" ({created_at})")
            };

            if url.is_empty() {
                lines.push(format!("- {prefix}{text}{meta}"));
            } else {
                lines.push(format!("- {prefix}{text}{meta} {url}"));
            }
        }

        if !qr.citations.is_empty() {
            lines.push(String::new());
            lines.push("**Citations:**".to_string());
            for c in &qr.citations {
                let url = c.url.as_deref().unwrap_or("");
                let title = c.title.as_deref().unwrap_or(url);
                if !url.is_empty() {
                    lines.push(format!("- [{title}]({url})"));
                }
            }
        }

        lines.push(String::new());
    }

    lines.join("\n").trim_end().to_string() + "\n"
}

fn load_existing_sources(workspace: &std::path::Path) -> HashSet<String> {
    let mut sources = HashSet::new();

    // Check candidates
    let cand_dir = workspace.join("memory").join("candidates");
    if cand_dir.exists() {
        if let Ok(entries) = std::fs::read_dir(&cand_dir) {
            for entry in entries.flatten() {
                let path = entry.path();
                if path.extension().and_then(|e| e.to_str()) == Some("jsonl") {
                    if let Ok(content) = std::fs::read_to_string(&path) {
                        for line in content.lines() {
                            let line = line.trim();
                            if line.is_empty() {
                                continue;
                            }
                            if let Ok(obj) = serde_json::from_str::<serde_json::Value>(line) {
                                if let Some(s) = obj.get("source").and_then(|v| v.as_str()) {
                                    sources.insert(s.to_string());
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    // Check store
    let store_dir = workspace.join("memory").join("store");
    for name in &["provisional.jsonl", "confirmed.jsonl", "rejected.jsonl"] {
        let p = store_dir.join(name);
        if let Ok(content) = std::fs::read_to_string(&p) {
            for line in content.lines() {
                let line = line.trim();
                if line.is_empty() {
                    continue;
                }
                if let Ok(obj) = serde_json::from_str::<serde_json::Value>(line) {
                    if let Some(s) = obj.get("source").and_then(|v| v.as_str()) {
                        sources.insert(s.to_string());
                    }
                }
            }
        }
    }

    sources
}
